File: credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int num_digits(long long n);
int sum_digits(int n);
bool luhn_pass(long long n);
bool amex_match(long long n);
bool mastercard_match(long long n);
bool visa_match(long long n);
string card_type(long long n);

int main(void)
{
    printf("Number: ");
    long long number = get_long_long();
    if (number > 0 && luhn_pass(number)) {
        printf("%s", card_type(number));
    }
    else {
        printf("INVALID\n");
    }
}

bool luhn_pass(long long n)
{
    int sub = 0;
    int even_dig = 0;
    while (n > 0) {
        if (even_dig == 0) {
            sub += n % 10;
            even_dig = 1;
        }
        else {
            sub += sum_digits((n % 10) * 2);
            even_dig = 0;
        }
        n = n / 10;
    }
    return sub % 10 == 0;
}

string card_type(long long n)
{
    if (amex_match(n)) {
        return ("AMEX\n");
    }
    else if (mastercard_match(n)) {
        return ("MASTERCARD\n");
    }
    else if (visa_match(n)) {
        return ("VISA\n");
    }
    else {
        return ("INVALID\n");
    }
}

int sum_digits(int n)
{
    int sub = 0;
    while (n > 0) {
        sub += n % 10;
        n = n / 10;
    }
    return sub;
}

int num_digits(long long n)
{
    return ceil(log10(n));
}

bool amex_match(long long n)
{
    return (num_digits(n) == 15 && (n / 10000000000000 == 34 || n / 10000000000000 == 37));
}

bool visa_match(long long n)
{
    return ((num_digits(n) == 13 || num_digits(n) == 16) && (n / 1000000000000000 == 4 || n / 1000000000000 == 4));
}

bool mastercard_match(long long n)
{
    return (num_digits(n) == 16 && (n / 100000000000000 == 51 || n / 100000000000000 == 52
                                       || n / 100000000000000 == 53 || n / 100000000000000 == 54 || n / 100000000000000 == 55));
}

File: credit.py
import math
import cs50


def main():
    # get credit card number from user
    number = cs50.get_int("Number: ")

    if number > 0 and luhn_pass(number):
        print(card_type(number))
    else:
        print("INVALID")
    # do I need this in Python version?
    return 0    


def luhn_pass(n):
    sub = 0
    even_dig = 0
    while(n > 0):
        if even_dig == 0:
            sub += n % 10
            even_dig = 1
        else:
            sub += sum_digits((n % 10) * 2)
            even_dig = 0
        n = n // 10    
    return sub % 10 == 0

    
def num_digits(n):
    return math.ceil(math.log(n,10))
    
    
def sum_digits(n):
    sub = 0
    while n > 0:
        sub += n % 10
        n = n //10
    return sub


def card_type(n):
    if amex_match(n):
        return "AMEX"
    elif mastercard_match(n):
        return "MASTERCARD"
    elif visa_match(n):
        return "VISA"
    else:
        return "INVALID"

    
def amex_match(n):
    return num_digits(n) == 15 and (n // 10000000000000 == 34 or n // 10000000000000 == 37)


def visa_match(n):
    return (num_digits(n) == 13 or num_digits(n) == 16) and (n // 1000000000000000 == 4 or n // 1000000000000 == 4)    
    

def mastercard_match(n):
    return num_digits(n) == 16 and (n // 100000000000000 == 51 or n // 100000000000000 == 52
                                       or n // 100000000000000 == 53 or n // 100000000000000 == 54 or n // 100000000000000 == 55)    
    
    
if __name__ == "__main__":
    main()

